Прохожу спринт 14

EntityVadim opened the pull request:
Дорогой ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
- Настроил проект под верстку кодом;
- Свертал кодом LaunchScreen;
- Сверстал кодом OnboardingViewController.

EntityVadim committed:
- Добавил иконки для верстки экранов;
- Делаю верстку кодом главного экрана.

EntityVadim committed:
- Добавил иконки для верстки кодом;
- Сверстал главный экран.

EntityVadim committed:
- Улучшил читаемость кода.

EntityVadim committed:
- Дописал верстку кодом OnboardingViewController.

EntityVadim committed:
- Улучшил читаемость кода.

EntityVadim committed:
- Сверстал экран статистики;
- Добавил иконку для верстки кодом.

EntityVadim committed:
- Внес мелкие коррективы по коду.

EntityVadim committed:
- Неправильно понял задание с датой и пришлось переделать код TrackersViewController.

EntityVadim committed:
- Пытаюсь исправить работы кнопки выбора даты.

EntityVadim committed:
- Не работают кнопки.

EntityVadim committed:
- Все верстку кодом переделываю, так как не добавил NavBarItem.

EntityVadim committed:
- Просил комит.

EntityVadim committed:
- Перераспределил файлы;
- Переписал верстку кодом.

EntityVadim committed:
- Пытаюсь поменять формат даты.

EntityVadim committed:
- Выполнил задание:
   Чек-лист по деталям реализации трекеров:
    - В проект добавлены структуры Tracker, TrackerCategory и TrackerRecord (они используются в UI части приложения).
    - Создан Tracker — сущность для хранения информации про трекер (для «Привычки» или «Нерегулярного события»). У
      него есть уникальный идентификатор (id), название, цвет, эмоджи и распиcание. Структуру данных для хранения
      расписания выберите на своё усмотрение.
    - Создан TrackerCategory — сущность для хранения трекеров по категориям. Она имеет заголовок и содержит массив
      трекеров, относящихся к этой категории.
    - TrackerRecord — сущность для хранения записи о том, что некий трекер был выполнен на некоторую дату; хранит id
      трекера, который был выполнен и дату. То есть, когда пользователь нажимает на + в ячейке трекера, мы создаём новую
      запись TrackerRecord — туда и запоминаем, какой трекер был выполнен в выбранную дату. Это может быть полезно,
      например, для того, чтобы отобразить в UI, был ли отмечен трекер в определенный день. Можно было бы просто хранить
      массив дат, в котором хранилась бы информация о выполненных датах, однако создание сущности позволит нам в
      будущем добавить новые поля и проводить различные дополнительные действиям с отметками. В качестве типа данных
      для id рекомендуем использовать тип, который однозначно определяет уникальный идентификатор — например, UInt или
      UUID. Не используйте такие типы, как Float или Date!
    - Структуры, Tracker, TrackerCategory и TrackerRecord сделаны без «мутабельных» (изменяемых) свойств. Это значит, что
      все их поля нужно пометить let. Мы помечаем поля let, т.к. на данный момент структуры используются только для
      передачи информации для отображения. В будущем будут добавлены сущности в базу данных и сгенерированы классы
      для работы с
      ними. На основе этих классов будут создаваться структуры для передачи данных.
    - Список категорий и вложенных в них трекеров хранится в свойстве var categories: [TrackerCategory] класса
      TrackersViewController.
    - Трекеры, которые были «выполнены» в выбранную дату, хранятся в var completedTrackers: [TrackerRecord]. Когда
      пользователь нажимает на + в ячейке трекера, добавляется соответствующая запись в completedTrackers. Если
      пользователь убирает пометку о выполненности в ячейке трекера, элемент удаляется из массива.
    - Новые трекеры добавляются в соответствующую категорию в массиве categories. Чтобы их добавить, нужно создать
      новую категорию с новым списком трекеров, а затем создать новый список категорий и присвоить его в categories. Мы не
      рекомендуем менять существующий массив, лучше создайте новый — так будет меньше пространства для
      трудноуловимых ошибок синхронизации данных.

EntityVadim committed:
- Добавил переход темы от светлой к темной, по состоянию настроек системы

EntityVadim committed:
- Выполняю задание:
1. Добавлен UICollectionView внутри TrackersViewController для отображения списка трекеров;
2. Свёрстана ячейка для трекера с кнопкой +. С помощью кнопки можно добавить запись, что этот трекер нужно пометить как выполненный для даты, выбранной в UIDatePicker.
3. Реализована логика отметки трекера выполненным:
- при нажатии на + изображение кнопки меняется на галочку «выполнено для указанной даты», а счётчик увеличивается;
- если трекер уже был выполнен для этой даты, то на кнопке появится галочка; при нажатии на галочку изображение поменяется обратно на +, а счётчик уменьшается;
- счётчик показывает количество отметок по карточке за всю историю;
- нельзя отметить карточку для будущей даты;

EntityVadim committed:
- Выполняю задание:
1. Добавлен UICollectionView внутри TrackersViewController для отображения списка трекеров;
2. Свёрстана ячейка для трекера с кнопкой +. С помощью кнопки можно добавить запись, что этот трекер нужно пометить как выполненный для даты, выбранной в UIDatePicker.
3. Реализована логика отметки трекера выполненным:
- при нажатии на + изображение кнопки меняется на галочку «выполнено для указанной даты», а счётчик увеличивается;
- если трекер уже был выполнен для этой даты, то на кнопке появится галочка; при нажатии на галочку изображение поменяется обратно на +, а счётчик уменьшается;
- счётчик показывает количество отметок по карточке за всю историю;
- нельзя отметить карточку для будущей даты.

EntityVadim committed:
- Выполняю задание:
1. Добавлен UICollectionView внутри TrackersViewController для отображения списка трекеров;
2. Свёрстана ячейка для трекера с кнопкой +. С помощью кнопки можно добавить запись, что этот трекер нужно пометить как выполненный для даты, выбранной в UIDatePicker.
3. Реализована логика отметки трекера выполненным:
- при нажатии на + изображение кнопки меняется на галочку «выполнено для указанной даты», а счётчик увеличивается;
- если трекер уже был выполнен для этой даты, то на кнопке появится галочка; при нажатии на галочку изображение поменяется обратно на +, а счётчик уменьшается;
- счётчик показывает количество отметок по карточке за всю историю;
- нельзя отметить карточку для будущей даты.

EntityVadim committed:
- Выполнил задание:
1. Добавлен UICollectionView внутри TrackersViewController для отображения списка трекеров;
2. Свёрстана ячейка для трекера с кнопкой +. С помощью кнопки можно добавить запись, что этот трекер нужно пометить как выполненный для даты, выбранной в UIDatePicker.
3. Реализована логика отметки трекера выполненным:
- при нажатии на + изображение кнопки меняется на галочку «выполнено для указанной даты», а счётчик увеличивается;
- если трекер уже был выполнен для этой даты, то на кнопке появится галочка; при нажатии на галочку изображение поменяется обратно на +, а счётчик уменьшается;
- счётчик показывает количество отметок по карточке за всю историю;
- нельзя отметить карточку для будущей даты.

EntityVadim committed:
- Сделал задание:
1. Реализованы экраны создания трекера:
 - при нажатии на кнопку + на главном экране показывается выбор типа трекера («Привычка» или «Нерегулярное событие»).

EntityVadim committed:
- Выполняю задание:
1. Реализованы экраны создания трекера:
 - при нажатии на кнопку + на главном экране показывается выбор типа трекера («Привычка» или «Нерегулярное событие»);
 - после выбора типа трекера пользователь попадает на экран создания трекера;
 - на экран создания трекера добавлены UITextField (для названия трекера) и кнопка перехода на задание расписания (если на предыдущем шаге было выбрано значение «Привычка»);
 - внизу экрана добавлены кнопки «Отменить» и «Сохранить». При нажатии на «Отменить» экран создания скрывается и не выполняются никакие изменения. При нажатии на «Сохранить» новый трекер добавляется в общий список и отображается на главном экране.

EntityVadim committed:
- Выполняю задание:
1. Реализованы экраны создания трекера:
- при нажатии на кнопку + на главном экране показывается выбор типа трекера («Привычка» или «Нерегулярное событие»);
- после выбора типа трекера пользователь попадает на экран создания трекера;
- на экран создания трекера добавлены UITextField (для названия трекера) и кнопка перехода на задание расписания (если на предыдущем шаге было выбрано значение «Привычка»);
- внизу экрана добавлены кнопки «Отменить» и «Сохранить». При нажатии на «Отменить» экран создания скрывается и не выполняются никакие изменения. При нажатии на «Сохранить» новый трекер добавляется в общий список и отображается на главном экране.

EntityVadim committed:
- Организовал код;
- Настроил кнопки по шаблону.

EntityVadim committed:
- Реализовал экран "Расписания" с сохранением данных.

EntityVadim committed:
- Добавил экран "Категория" и настроил его по магету.

EntityVadim committed:
- Добавил и настроил экран "Новая категория";
- Мелкие корректировки по коду.

EntityVadim committed:
- Добавил таблицу в TrackerCategoryViewController.

EntityVadim committed:
- Добавил возможность сохранять категории;
- Добавил возможность выбора категории.

EntityVadim committed:
- Создал кастомную галочку для ячейке в таблице для выбора категории.

EntityVadim committed:
- Добавил возможно при нажатии на кнопку "Нерегулярные события" скрывать кнопку "Расписание".

EntityVadim committed:
- Обновил код коллекции в TrackersViewController;
- Дал нормальные названия методам;
- Обновил хранения рекорда.

EntityVadim committed:
- Исправил ошибку из-за которой новый трекер не отображался на главном экране;
- Добавил возможность добавлять свои категории.

EntityVadim committed:
- Устранил мелкую ошибку.

EntityVadim committed:
- Сделал ползунки синими при выборе даты;
- Настроил таблицу по макету.

EntityVadim committed:
- Оставил комментарий для Ревью.

EntityVadim committed:
- Создал хранилище данных;
- Переименовал главный контроллер;
- Написал код для названия категорий и настроил его по макету;
- Связал код с хранением данных.

EntityVadim committed:
- Настроил по макету коллекцию;
- Исправил проблему передачи данных в коллекциюю;
- Переорганизовал код.

EntityVadim committed:
- Переорганизовал код.

EntityVadim committed:
- Обновил размер плюса и галочки при выполнения трекера;
- Добавил состояния кнопки при выполнения трекера;
- Добавил увеличения счетчика выполненых трекеров в разные даты.

EntityVadim committed:
- Доделал главный экран по макету.

Helvetii951 commented on Tracker/AppDelegate.swift:
```diff
@@ -9,9 +9,21 @@ import UIKit
+                let rootViewController = MainTabBarController()
+                window?.rootViewController = rootViewController
+            }
+            window?.makeKeyAndVisible()
```
Для хранения подобных флагов стоит создать отдельный класс. С ростом приложения настроек и флагов будет больше. Хранить их в одном месте будет очень удобно.

Кстати, ты чуть поторопился с онбордингом, он будет в 16м спринте :)

Helvetii951 commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,153 @@
+
+import UIKit
+
+final class OnboardingViewController: UIViewController {
+    
+    // MARK: - Private Properties
+    
+    private lazy var scrollView: UIScrollView = {
+        let scrollView = UIScrollView()
+        scrollView.isPagingEnabled = true
+        scrollView.showsHorizontalScrollIndicator = false
+        return scrollView
+    }()
```
В 16 спринте предполагается, что "листалка" будет реализована с помощью [UIPageViewController](https://developer.apple.com/documentation/uikit/uipageviewcontroller). Возможно, придется переделать.

Helvetii951 commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,153 @@
+        return view
+    }()
+    
+    private lazy var firstLabel: UILabel = {
+        return makeLabel(withText: "Отслеживайте только то, что хотите")
+    }()
+    
+    private lazy var secondLabel: UILabel = {
+        return makeLabel(withText: "Даже если это не литры воды и йога")
+    }()
+    
+    private lazy var firstActionButton: UIButton = {
+        return makeActionButton(withTitle: "Вот это технологии!")
+    }()
+    
+    private lazy var secondActionButton: UIButton = {
+        return makeActionButton(withTitle: "Вот это технологии!")
+    }()
+    
+    private lazy var firstScreenView: UIImageView = {
+        return makeImageView(withImageNamed: "OnboardingBlue")
+    }()
+    
+    private lazy var secondScreenView: UIImageView = {
+        return makeImageView(withImageNamed: "OnboardingRed")
+    }()
```
Чтобы сократить код и сделать его более гибким, рекомендую для отображения страницы онбординга создать отдельный подкласс `UIView` или `UIViewController` , в котором UI компоненты создаются и настраиваются уже внутри. Так главному контроллеру всего лишь нужно будет создать экземпляр этого подкласса и передать картинку+текст.

Helvetii951 replied:
Было бы классно обернуть инфу о странице(текст и картинку) в enum. [Сделать эту инфу вычислимыми свойствами](https://www.avanderlee.com/swift/computed-property/). Так код будет гибче.

Helvetii951 commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,153 @@
+        return scrollView
+    }()
+    
+    private lazy var contentView: UIView = {
+        let view = UIView()
+        return view
+    }()
```
В онбординге не хватает индикатора текущей страницы. Чтобы определить текущую страницу, придется задействовать [UIScrollViewDelegate](https://developer.apple.com/documentation/uikit/uiscrollviewdelegate) примерно так:
```swift
func scrollViewDidScroll(_ scrollView: UIScrollView) {
    let pageIndex = round(scrollView.contentOffset.x / view.frame.width)
    // pageControl - экземпляр UIPageControl
    pageControl.currentPage = Int(pageIndex)
}
```

Helvetii951 commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,153 @@
+        return imageView
+    }
+    
+    @objc private func didTapActionButton() {
+        UserDefaults.standard.set(true, forKey: "isOnboardingCompleted")
+        let mainTabBarController = MainTabBarController()
+        if let window = UIApplication.shared.windows.first {
+            window.rootViewController = mainTabBarController
+            UIView.transition(
+                with: window,
+                duration: 0.5,
+                options: .transitionCrossDissolve,
+                animations: nil,
+                completion: nil)
+        }
+    }
```
Лучше забрать у контроллера онбординга обязанность навигации и отдать. ее тому, кто создал контроллер(в нашем случае это `SceneDelegate`). Передаем событие нажатия кнопки через, например, замыкание, а `SceneDelegate` уже пусть сам решает что делать.

Helvetii951 commented on Tracker/Trackers/TrackersDataStorage/TrackerModels.swift:
```diff
@@ -0,0 +1,26 @@
+
+import UIKit
+
+struct Tracker {
+    let id: UUID
+    let name: String
+    let color: UIColor
+    let emoji: String
+    let schedule: [String]
+}
+
+struct TrackerCategory {
+    let title: String
+    let trackers: [Tracker]
+}
+
+struct TrackerRecord: Hashable {
+    let trackerId: UUID
+    let date: String
+}
```
Рекомендую раскидывать такие модели по отдельным файлам. Так навигация по проекту будет удобнее.

Helvetii951 commented on Tracker/Trackers/TrackerCategoryViewController.swift:
```diff
@@ -0,0 +1,202 @@
+        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "CategoryCell")
+        tableView.separatorStyle = .singleLine
+        tableView.separatorColor = UIColor.ypGrey
+        tableView.separatorInset = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
+        tableView.rowHeight = 60 // ❗❗❗Не понимаю, почему не работает❗❗❗
+        return tableView
```
Работает :)
<img width="1006" alt="Снимок экрана 2024-08-11 в 07 29 04" src="https://github.com/user-attachments/assets/8eba1a2e-0083-44a9-9003-4f8ea773c5b3">

Helvetii951 commented on Tracker/Trackers/TrackerCollection/TrackerSectionHeader.swift:
```diff
@@ -0,0 +1,37 @@
+
+final class TrackerSectionHeader: UICollectionReusableView {
+    
+    let titleLabel: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.systemFont(ofSize: 19, weight: .bold)
+        label.textColor = .ypBlack
+        return label
+    }()
```
Рекомендую заприватить вьюху, настраивать ее через специальную функцию:
```
func configHeader(with title : String)
```

Так код будет гибче. Если мы, например, поменяем внтреннюю реализацию UI, нам не придется переписывать код в контоллере.

Helvetii951 commented on Tracker/Trackers/TrackerViewController.swift:
```diff
@@ -0,0 +1,284 @@
+            return CGSize(width: collectionView.frame.width, height: 30)
+        }
+    
+    func collectionView(
+        _ collectionView: UICollectionView,
+        layout collectionViewLayout: UICollectionViewLayout,
+        sizeForItemAt indexPath: IndexPath
+    ) -> CGSize {
+        return CGSize(width: 167, height: 148)
+    }
```
Принято считать размер(в нашем случае ширину) динамически в зависимости от размеров `UICollectioView` и внутренних отступов. Как-то так:
```swift
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        // 20 - это сумма отступов в одном из проектов
        let cellWidth = collectionView.bounds.width / 2 - 20
        return CGSize(width: cellWidth, height: 148)
    }
```

EntityVadim committed:
- Цвет активного таба поменял с черного на синий;
- Сверстал приложение по макету;
- Стартовый контроллер установил в SceneDelegate.

EntityVadim committed:
- Обновил отображение категории и дни недели по макету.

EntityVadim committed:
- Разделил модель на разные файлы для лучшей гибкости кода.

EntityVadim committed:
- Организовал код для лучшей читаемости.

EntityVadim committed:
- Сделал возможность блокировки кнопки, если данные по созданию новой привычки не заполнены и фон кнопки серы, если всу условия выполнены, то кнопка разблокируется и фон становится черным.

EntityVadim committed:
- Исправил мелкие ошибки по коду (экран создания привычки не соответствовал макету.

EntityVadim committed:
- Сделал фильтрацию по датам;
- Ограничил возможность отмечать трекер в будущем.

EntityVadim committed:
- Потребовал комит

EntityVadim committed:
- Добавил возможность отображение трекера в определенный день.

EntityVadim committed:
- Обвновил код коллекции, которая не понимала, как сортировать трекеры.

EntityVadim committed:
- Обновил код с учетом этого требования:
Напомним, что Привычка отличается от Нерегулярного события тем, что нерегулярное событие выполняется только один раз и не переходит в другие дни, если оно было выполнено. Привычка же отображается на списках трекеров согласно ее расписанию. Например, если создать привычку с названием "Занятие спортом" и указать ей расписание 3 раза в неделю - понедельник, среда и пятница, тогда выбирая эти дни недели, привычка должна отображаться в списке трекеров.

EntityVadim committed:
- Обновил код для трекера с типо "Нерегулярное событие" (Должен отображаться завершенным и не пропадать)

EntityVadim committed:
- Пытаюсь исправить ошибку из-за которой трекер не удаляется.

EntityVadim committed:
- Исправил баг с "Нерегулярное событие"
